docs/guides/transfer_learning.ipynb

```diff
@@ -0,0 +1,308 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": [
+     "skip-execution"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "! pip install transformers[flax]\n",
+    "! pip install -U jax jaxlib"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating the model\n",
+    "We will start by creating a `load_model` function that will return a Flax Module and its pretrained variables. In this specific example we'll be using HuggingFace's `FlaxCLIPVisionModel` model, which will extract a `FlaxCLIPModule` module."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "from IPython.display import clear_output\n",
+    "from transformers import FlaxCLIPModel\n",
+    "\n",
+    "# Note: FlaxCLIPModel is not a Flax Module\n",
+    "def load_model():\n",
+    "  clip = FlaxCLIPModel.from_pretrained('openai/clip-vit-base-patch32')\n",
+    "  clear_output(wait=False) # clear the loading messages\n",
+    "  module = clip.module # extract the Flax Module\n",
+    "  variables = {'params': clip.params}\n",
+    "  return module, variables"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Extracting a submodule\n",
+    "Calling `load_model` gives us the `FlaxCLIPModule` which is composed of a text and vision submodules. Let suppose we want the `vision_model` submodule defined inside `.setup()`, to do this we we can use `bind` to force `.setup()` to run (see [The Module lifecycle](https://flax.readthedocs.io/en/latest/advanced_topics/module_lifecycle.html#setup)) and grant us access to submodules and their variables."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "clip_module, clip_variables = load_model()\n",
+    "# bind variables, this runs .setup() making .vision_model available\n",
+    "bounded_clip = clip_module.bind(clip_variables)\n",
+    "# returned an unbounded copy and corresponding variables\n",
+    "vision_model = bounded_clip.vision_model.clone()\n",
+    "vision_model_variables = bounded_clip.vision_model.variables"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that we used `.clone()` to get an unbounded copy of `vision_model` and `.variables` to get the correct variable structure.\n",
+    "\n",
+    "### Creating a new model\n",
+    "Next we will create a `Classifier` module which will consist on a `backbone` (the pretrained vision model) and a `head` (the classifier)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import flax.linen as nn\n",
+    "import jax.numpy as jnp\n",
+    "import jax\n",
+    "\n",
+    "class Classifier(nn.Module):\n",
+    "  num_classes: int\n",
+    "  backbone: nn.Module\n",
```

Comment:
Perhaps we want to use this as an example but we should probably warn users that this is a particularly bad inversion of control. You lose the ability to cache the features resulting from backbone which is typically the majority of the compute. And it also requires complicated surgery both to fine-tune the classifier or end2end.

I notice that many users intuitively code something like this and then get bitten by the complexity it introduces so we should really highlight the simple alternative that fits >90% of the use cases IMO

Reply:
This was "solved" via a note after the introduction.